### PR TITLE
[reland] Report test time regressions

### PR DESCRIPTION
**Pull request by samestep**
This is a followup to #49190. Vaguely speaking, the goals are to make it easy to identify test time regressions introduced by PRs. Eventually the hope is to use this information to edit Dr CI comments, but this particular PR just does the analysis and prints it to stdout, so a followup PR would be needed to edit the actual comments on GitHub.

**Important:** for uninteresting reasons, this PR moves the `print_test_stats.py` file.

- *Before:* `test/print_test_stats.py`
- *After:* `torch/testing/_internal/print_test_stats.py`

Notes on the approach:

- Just getting the mean and stdev for the total job time of the last _N_ commits isn't sufficient, because e.g. if `master` was broken 5 commits ago, then a lot of those job times will be much shorter, breaking the statistics.
- We use the commit history to make better estimates for the mean and stdev of individual test (and suite) times, but only when the test in that historical commit is present and its status matches that of the base commit.
- We list all the tests that were removed or added, or whose status changed (e.g. skipped to not skipped, or vice versa), along with time (estimate) info for that test case and its containing suite.
- We don't list tests whose time changed a lot if their status didn't change, because there's a lot of noise and it's unclear how to do that well without too many false positives.
- We show a human-readable commit graph that indicates exactly how many commits are in the pool of commits that could be causing regressions (e.g. if a PR has multiple commits in it, or if the base commit on `master` doesn't have a report in S3).
- We don't show an overall estimate of whether the PR increased or decreased the total test job time, because it's noisy and it's a bit tricky to aggregate stdevs up from individual tests to the whole job level. This might change in a followup PR.
- Instead, we simply show a summary at the bottom which says how many tests were removed/added/modified (where "modified" means that the status changed), and our best estimates of the mean times (and stdevs) of those changes.
- Importantly, the summary at the bottom is only for the test cases that were already shown in the more verbose diff report, and does not include any information about tests whose status didn't change but whose running time got much longer.

**Test Plan:**

To run the unit tests:
```
$ python test/test_testing.py
$ python test/print_test_stats.py
```

To verify that this works, check the [CircleCI logs](https://app.circleci.com/pipelines/github/pytorch/pytorch/258628/workflows/9cfadc34-e042-485e-b3b3-dc251f160307) for a test job run on this PR; for example:
- pytorch_linux_bionic_py3_6_clang9_test

To test locally, use the following steps.

First run an arbitrary test suite (you need to have some XML reports so that `test/print_test_stats.py` runs, but we'll be ignoring them here via the `--use-json` CLI option):
```
$ DATA_DIR=/tmp
$ ARBITRARY_TEST=testing
$ python test/test_$ARBITRARY_TEST.py --save-xml=$DATA_DIR/test/test_$ARBITRARY_TEST
```
Now choose a commit and a test job (it has to be on `master` since we're going to grab the test time data from S3, and [we only upload test times to S3 on the `master`, `nightly`, and `release` branches](https://github.com/pytorch/pytorch/pull/49645)):
```
$ export CIRCLE_SHA1=c39fb9771d89632c5c3a163d3c00af3bef1bd489
$ export CIRCLE_JOB=pytorch_linux_bionic_py3_6_clang9_test
```
Download the `*.json.bz2` file(s) for that commit/job pair:
```
$ aws s3 cp s3://ossci-metrics/test_time/$CIRCLE_SHA1/$CIRCLE_JOB/ $DATA_DIR/ossci-metrics/test_time/$CIRCLE_SHA1/$CIRCLE_JOB --recursive
```
And feed everything into `test/print_test_stats.py`:
```
$ bzip2 -kdc $DATA_DIR/ossci-metrics/test_time/$CIRCLE_SHA1/$CIRCLE_JOB/*Z.json.bz2 | torch/testing/_internal/print_test_stats.py --compare-with-s3 --use-json=/dev/stdin $DATA_DIR/test/test_$ARBITRARY_TEST
```
The first part of the output should be the same as before this PR; here is the new part, at the end of the output:

- https://pastebin.com/Jj1svhAn

### CODE REVIEW

File: test/print_test_stats.py [hunk 1]
```diff
@@ -156,26 +183,67 @@ def send_report_to_s3(reports, *, total_seconds):
...
-                    }
-                    for case in suite.test_cases
-                ],
-            }
-            for name, suite in reports.items()
-        }
-    }).encode()))
+    obj.put(Body=bz2.compress(json.dumps(obj).encode()))
+
+def print_regressions(obj, *, n, stdev_threshold):
```
Comment (walterddr) [inline test/print_test_stats.py:188]: `n` is a vague variable name

File: test/print_test_stats.py [hunk 2]
```diff
@@ -156,26 +183,67 @@ def send_report_to_s3(reports, *, total_seconds):
...
+    objects = {}
+    # should we do these in parallel?
+    for commit, summaries in index.items():
+        objects[commit] = []
+        for summary in summaries:
+            binary = summary.get()["Body"].read()
+            string = bz2.decompress(binary).decode("utf-8")
+            objects[commit].append(json.loads(string))
+
+    print()
```
Comment (walterddr) [inline test/print_test_stats.py:236]: is this just for a line separator?

File: test/print_test_stats.py (hunk 1 above)
Comment (walterddr) [reply test/print_test_stats.py:188]: also nit: the major thing this function does seems like cleaning up the json object to split it into head and base, yes? can we break that out into a separate function --> figure that could be useful elsewhere too.

Comment (samestep) [reply test/print_test_stats.py:188]: > `n` is a vague variable name

good point, will change that

> also nit: the major thing this function does seems like cleaning up the json object to split it into head and base, yes? can we break that out into a separate function --> figure that could be useful elsewhere too.

not sure exactly what you mean here; the base info isn't in the JSON object it's passed (`obj`), but rather in S3, so it needs to run `git` commands to figure out what commits to grab from S3, then download those, and then pass to `stats_utils`

File: test/print_test_stats.py (hunk 2 above)
Comment (samestep) [reply test/print_test_stats.py:236]: yep

File: test/print_test_stats.py (hunk 1 above)
Comment (walterddr) [reply test/print_test_stats.py:188]: Yes. What I meant was spin off into a `(head, base) = fetch_base_content(obj, n)` 
Then in print function it does fetch, then calls stats_utils and get human-readable format.

File: test/print_test_stats.py (hunk 2 above)
Comment (walterddr) [reply test/print_test_stats.py:236]: Ah. Cool. Could you add msg like `----- Historic stats comparison result ------`

Comment (samestep) [reply test/print_test_stats.py:236]: There is already a message (shown in the pastebin); it looks like this:

```
Following output is to check for test time regressions:
    job: <job>
    commit: <sha1>
```

Do you want me to change it (e.g. by adding more hyphens)?

File: test/print_test_stats.py (hunk 1 above)
Comment (samestep) [reply test/print_test_stats.py]: so you're saying, add a function that computes the 7 arguments to pass to `stats_utils.regression_info`, then call that function, extract the 7 arguments, and pass the results to `stats_utils.regression_info`?

File: test/print_test_stats.py (hunk 2 above)
Comment (walterddr) [reply test/print_test_stats.py:236]: no it is fine

File: test/print_test_stats.py (hunk 1 above)
Comment (walterddr) [reply test/print_test_stats.py]: nvm. let's do this in follow up PR

**Review by walterddr — APPROVED** [review]

**Comment by samestep** [comment]
Reverting because this seems to have broken multiple jobs:

- [pytorch_linux_xenial_cuda10_2_cudnn7_py3_jit_legacy_test](https://app.circleci.com/pipelines/github/pytorch/pytorch/270965/workflows/834de0e1-9633-42ca-8e49-c0185e3156d2/jobs/10758342)
- [pytorch_linux_xenial_cuda10_2_cudnn7_py3_multigpu_test](https://app.circleci.com/pipelines/github/pytorch/pytorch/270965/workflows/834de0e1-9633-42ca-8e49-c0185e3156d2/jobs/10758344)
- [pytorch_linux_xenial_py3_6_gcc5_4_jit_legacy_test](https://app.circleci.com/pipelines/github/pytorch/pytorch/270965/workflows/834de0e1-9633-42ca-8e49-c0185e3156d2/jobs/10757495)
- [pytorch_linux_xenial_py3_clang7_onnx_ort_test1](https://app.circleci.com/pipelines/github/pytorch/pytorch/270965/workflows/834de0e1-9633-42ca-8e49-c0185e3156d2/jobs/10757328)
- [pytorch_linux_xenial_py3_clang7_onnx_ort_test2](https://app.circleci.com/pipelines/github/pytorch/pytorch/270965/workflows/834de0e1-9633-42ca-8e49-c0185e3156d2/jobs/10757313)

**Pull request closed without merging**